### PR TITLE
[RFC/WIP] Leverage compiler-rt for builtins support of llvm intrinsics

### PR DESCRIPTION
**Pull request by vchuravy**
## Note:

This feature will remain unavailable until we compile LLVM with cmake and bump our llvm version to ~~3.8~~ 3.9
### Reason

llvm will only build `compiler-rt` with cmake and turns building it with autotools explicitly off. We currently build llvm as a shlib and that only works in combination with cmake from version 3.8 upwards.

[Compiler-rt](http://compiler-rt.llvm.org/) (builtins) provides low-level support for llvm intrinsics that don't map to an appropriate cpu instructions.

> builtins - a simple library that provides an implementation of the low-level target-specific hooks required by code generation and other runtime components. For example, when compiling for a 32-bit target, converting a double to a 64-bit unsigned integer is compiling into a runtime call to the "__fixunsdfdi" function. The builtins library provides optimized implementations of this and other low-level routines, either in target-independent C form, or as a heavily-optimized assembly.
> builtins provides full support for the libgcc interfaces on supported targets and high performance hand tuned implementations of commonly used functions like __floatundidf in assembly that are dramatically faster than the libgcc implementations. It should be very easy to bring builtins to support a new target by adding the new routines needed by that target.

It provides fallbacks for llvm intrinsics on platfroms, where they don't map to an appropriate instruction. As an example we could use the cpu instructions to convert Float16 -> Float32 on any platform that supports `f16c`.
### Caveats
- Compiler-rt is build and distributed as a static library so we have to convert it into a shared library.
- Tested under Linux x86_64 and the build system will need adjustments for other targets.
- compiler-rt build system is quite special (see rusts build-system https://github.com/rust-lang/rust/blob/master/mk/rt.mk)
- needs cmake and llvm 3.8
### Todo
- [ ] Buildsystem support for Windows/ARM/PPC/Mac
- ~~ Enable building with `LLVM_USE_CMAKE = 0`~~
- [x] Support `USE_SYSTEM_LLVM = 1`
- [x] Build per default

### CODE REVIEW

**Comment by maleadt** [comment]
Some timings on `Float16` conversions, on current master:
```
julia> @benchmark Float16(1.0f0)
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     4.304 ns (0.00% GC)
  median time:      4.312 ns (0.00% GC)
  mean time:        4.342 ns (0.00% GC)
  maximum time:     7.633 ns (0.00% GC)

julia> @benchmark Float16(1.0)
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     4.306 ns (0.00% GC)
  median time:      4.313 ns (0.00% GC)
  mean time:        4.345 ns (0.00% GC)
  maximum time:     8.068 ns (0.00% GC)

julia> @benchmark Float32(Float16(1.0))
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     6.799 ns (0.00% GC)
  median time:      6.808 ns (0.00% GC)
  mean time:        7.015 ns (0.00% GC)
  maximum time:     12.528 ns (0.00% GC)

julia> @benchmark Float64(Float16(1.0))
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     6.787 ns (0.00% GC)
  median time:      6.792 ns (0.00% GC)
  mean time:        6.803 ns (0.00% GC)
  maximum time:     10.169 ns (0.00% GC)

julia> @benchmark Float16(1.0) + 1
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     6.799 ns (0.00% GC)
  median time:      6.808 ns (0.00% GC)
  mean time:        7.152 ns (0.00% GC)
  maximum time:     9.103 ns (0.00% GC)
```

And with this PR:
```
julia> @benchmark Float16(1.0f0)
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     1.368 ns (0.00% GC)
  median time:      1.371 ns (0.00% GC)
  mean time:        1.373 ns (0.00% GC)
  maximum time:     3.727 ns (0.00% GC)

julia> @benchmark Float16(1.0)
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     1.370 ns (0.00% GC)
  median time:      1.373 ns (0.00% GC)
  mean time:        1.375 ns (0.00% GC)
  maximum time:     4.262 ns (0.00% GC)

julia> @benchmark Float32(Float16(1.0))
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     1.370 ns (0.00% GC)
  median time:      1.373 ns (0.00% GC)
  mean time:        1.375 ns (0.00% GC)
  maximum time:     2.876 ns (0.00% GC)

julia> @benchmark Float64(Float16(1.0))
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     1.369 ns (0.00% GC)
  median time:      1.372 ns (0.00% GC)
  mean time:        1.374 ns (0.00% GC)
  maximum time:     3.667 ns (0.00% GC)

julia> @benchmark Float16(1.0) + 1
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     1.367 ns (0.00% GC)
  median time:      1.371 ns (0.00% GC)
  mean time:        1.374 ns (0.00% GC)
  maximum time:     4.347 ns (0.00% GC)
```

File: base/float.jl [hunk 1]
```diff
@@ -14,7 +14,7 @@ const NaN = NaN64
 ## conversions to floating-point ##
 convert(::Type{Float16}, x::Integer) = convert(Float16, convert(Float32,x))
 for t in (Int8,Int16,Int32,Int64,Int128,UInt8,UInt16,UInt32,UInt64,UInt128)
-    @eval promote_rule(::Type{Float16}, ::Type{$t}) = Float32
+    @eval promote_rule(::Type{Float16}, ::Type{$t}) = Float16
```
Comment (yuyichao) [inline base/float.jl]: This should probably be a different PR

Comment (vchuravy) [inline base/float.jl]: Yes I should have noted that this is https://github.com/JuliaLang/julia/pull/17297

File: Make.inc [hunk 2]
```diff
@@ -42,6 +42,9 @@ USE_SYSTEM_PATCHELF:=0
 # Link to the LLVM shared library
 USE_LLVM_SHLIB := 1
 
+# Use the builtins from compiler-rt
+USE_COMPILER-RT :=0
```
Comment (tkelman) [inline Make.inc]: I'd be inclined to use underscores in variable names rather than dashes, for consistency's sake

File: Makefile [hunk 3]
```diff
@@ -180,7 +180,8 @@ endif
 $(build_private_libdir)/%.$(SHLIB_EXT): $(build_private_libdir)/%.o
 	@$(call PRINT_LINK, $(CXX) $(LDFLAGS) -shared $(fPIC) -L$(build_private_libdir) -L$(build_libdir) -L$(build_shlibdir) -o $@ $< \
 		$(if $(findstring -debug.$(SHLIB_EXT),$(notdir $@)),-ljulia-debug,-ljulia) \
-		$$([ $(OS) = WINNT ] && echo '' -lssp))
+		$$([ $(OS) = WINNT ] && echo '' -lssp) \
+		$$([ $(BUILD_COMPILER-RT) = 1 ] && echo '' -lcompiler-rt))
```
Comment (tkelman) [inline Makefile]: shouldn't llvm-config be doing this?

File: base/float16.jl [hunk 4]
```diff
@@ -1,102 +1,29 @@
...
-            (f & ((1<<(sh-1))-1)) != 0  # check lower bits
-            h += 1
-        end
-    end
-    reinterpret(Float16, h)
-end
+import Base.llvmcall
+# Implement conversion to and from Float16 with llvm intrinsics
+convert(::Type{Float32}, val::Float16) =
+  llvmcall(("""declare float @llvm.convert.from.fp16.f32(i16)""",
```
Comment (tkelman) [inline base/float16.jl]: 4-space indent

File: deps/llvm.mk [hunk 5]
```diff
@@ -454,11 +475,11 @@ endif # LLVM_VER
 ifeq ($(LLVM_VER),3.7.1)
 ifeq ($(BUILD_LLDB),1)
 $(eval $(call LLVM_PATCH,lldb-3.7.1))
-endif
-ifeq ($(BUILD_LLVM_CLANG),1)
+ifeq ($(BUILD_COMPILER-RT),1)
```
Comment (tkelman) [inline deps/llvm.mk]: should this patch get applied even when `BUILD_LLDB` is 0?

File: deps/llvm.mk [hunk 6]
```diff
@@ -512,6 +533,9 @@ else
 	$(call   make-install,llvm-$(LLVM_VER)/build_$(LLVM_BUILDTYPE),\
 	                      $(LLVM_MFLAGS) PATH="$(llvm_python_workaround):$$PATH")
 endif # LLVM_USE_CMAKE
+ifeq ($(BUILD_COMPILER-RT),1)
+	$(CC) $(LDFLAGS) -shared $(fPIC) -o $(build_libdir)/libcompiler-rt.$(SHLIB_EXT) -nostdlib -Wl,--whole-archive -L$(build_libdir)/clang/$(LLVM_VER)/lib/$(call patsubst,%inux,linux,$(OS)) -lclang_rt.builtins-$(call patsubst,i%86,i386,$(ARCH))
```
Comment (tkelman) [inline deps/llvm.mk]: probably should be `build_shlibdir`

Comment (vchuravy) [inline deps/llvm.mk]: I was even considering making this build_private_libdir, since compiler-rt is distributed as a static library, but we need a shared one to properly find the symbols at runtime.

**Comment by vchuravy** [comment]
@tkelman thanks for the review and I hope I addressed most of your comments.

One point that I would like to discuss is if we want to put the shared library version of compiler-rt to go into `build_private_libdir`, since we create it purely to be able for Julia to resolve symbols to it at runtime and there is no other consumer (clang and rust link to the static version) and we need to distribute it ourselves.

**Comment by tkelman** [comment]
Why would we put it in `build_private_libdir`? That's not really on the search path, from what I can tell the only thing we put there is the system image.

**Comment by vchuravy** [comment]
The place I would think is appropriate is lib/julia to emphasise that we
are the only consumer of that shared library.

**Comment by tkelman** [comment]
Shared libraries will be put in `lib/julia` in the install tree on unix.

**Comment by vchuravy** [comment]
I will leave this branch up for now, but I currently don't see a way of doing this properly (and not lose shlib support), without switching to 3.9

File: deps/Versions.make [hunk 7]
```diff
@@ -1,4 +1,4 @@
-LLVM_VER = 3.7.1
+LLVM_VER = 3.8.0
```
Comment (tkelman) [inline deps/Versions.make]: we'll probably do 3.8.1 some time not too long after branching for 0.6-dev, but it'll need CI preparation

**Comment by vchuravy** [comment]
For the build-system we could try to use the Makefile from https://github.com/ReservedField/arm-compiler-rt/ to make your lifes easier (not using cmake)

**Comment by vchuravy** [comment]
#18734 has the basic support for compiler-rt and I plan to add PR for full Float16 support after that got merged
